File: CSVExplorerSwift/CSVExplorerSwift/service/UploadClient.swift
import Foundation

final class UploadClient {

    static let endpoint = URL(string: "https://alperensaracdeneme.com/deneme/upload_csv.php")!

    struct UploadResponse: Decodable {
        let ok: Bool
        let id: String?
        let download_url: String?
        let error: String?
    }

    // ✅ PHP: $_FILES['file'] bekliyor -> fieldName = "file"
    static func uploadCsv(fileUrl: URL) async throws -> URL {
        var req = URLRequest(url: endpoint)
        req.httpMethod = "POST"

        let boundary = "Boundary-\(UUID().uuidString)"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        let fileData = try Data(contentsOf: fileUrl)
        let filename = fileUrl.lastPathComponent

        var body = Data()
        func append(_ s: String) { body.append(Data(s.utf8)) }

        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\n")
        append("Content-Type: text/csv\r\n\r\n")
        body.append(fileData)
        append("\r\n")
        append("--\(boundary)--\r\n")

        req.httpBody = body

        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else {
            throw NSError(domain: "upload", code: -1, userInfo: [NSLocalizedDescriptionKey: "No HTTP response"])
        }

        // Sunucu JSON dönüyor; hata da JSON ile geliyor
        let decoded = (try? JSONDecoder().decode(UploadResponse.self, from: data))
        if !(200..<300).contains(http.statusCode) {
            let msg = decoded?.error ?? (String(data: data, encoding: .utf8) ?? "Upload failed")
            throw NSError(domain: "upload", code: http.statusCode, userInfo: [NSLocalizedDescriptionKey: msg])
        }

        guard let res = decoded, res.ok == true else {
            let msg = decoded?.error ?? "Unknown error"
            throw NSError(domain: "upload", code: -2, userInfo: [NSLocalizedDescriptionKey: msg])
        }

        guard let urlStr = res.download_url, let url = URL(string: urlStr) else {
            throw NSError(domain: "upload", code: -3, userInfo: [NSLocalizedDescriptionKey: "download_url missing"])
        }

        return url
    }
}

File: csv_explorer_react_native/src/domain/upload_client.ts
import axios from "axios";

type UploadResponse =
    | string
    | { download_url?: string; url?: string; file_url?: string; link?: string; [k: string]: unknown };

export async function uploadCsv(endpoint: string, uri: string): Promise<string> {
    const form = new FormData();

    // PHP genelde "file" bekler. Eğer endpoint $_FILES['csv'] istiyorsa: "file" -> "csv"
    form.append("file", {
        uri,
        name: `upload_${Date.now()}.csv`,
        type: "text/csv",
    } as any);

    const res = await axios.post<UploadResponse>(endpoint, form, {
        headers: { "Content-Type": "multipart/form-data" },
        timeout: 60000,
    });

    const body = res.data;

    if (body && typeof body === "object") {
        const url = (body.download_url || body.url || body.file_url || body.link || "").toString().trim();
        if (url) return url;
    }

    const text = typeof body === "string" ? body : JSON.stringify(body);
    const m = text.match(/https?:\/\/\S+/);
    if (m?.[0]) return m[0];

    throw new Error("No download URL found in server response.");
}
